be/src/olap/rowset/beta_rowset_writer.cpp

```diff
@@ -0,0 +1,184 @@
+    }
+    _rowset_meta->set_tablet_uid(_context.tablet_uid);
+
+    return OLAP_SUCCESS;
+}
+
+template<typename RowType>
+OLAPStatus BetaRowsetWriter::_add_row(const RowType& row) {
+    if (PREDICT_FALSE(_segment_writer == nullptr)) {
+        RETURN_NOT_OK(_create_segment_writer());
+    }
+    // TODO update rowset's zonemap
+    auto s = _segment_writer->append_row(row);
+    if (PREDICT_FALSE(!s.ok())) {
+        LOG(WARNING) << "failed to append row: " << s.to_string();
+        return OLAP_ERR_WRITER_DATA_WRITE_ERROR;
+    }
+    if (PREDICT_FALSE(_segment_writer->estimate_segment_size() >= _max_segment_size)) {
+        RETURN_NOT_OK(_flush_segment_writer());
+    }
+    _num_rows_written++;
+    return OLAP_SUCCESS;
+}
+
+template OLAPStatus BetaRowsetWriter::_add_row(const RowCursor& row);
+template OLAPStatus BetaRowsetWriter::_add_row(const ContiguousRow& row);
+
+OLAPStatus BetaRowsetWriter::add_rowset(RowsetSharedPtr rowset) {
+    assert(rowset->rowset_meta()->rowset_type() == BETA_ROWSET);
+    RETURN_NOT_OK(rowset->link_files_to(_context.rowset_path_prefix, _context.rowset_id));
+    _num_rows_written += rowset->num_rows();
+    _total_data_size += rowset->rowset_meta()->data_disk_size();
+    _num_segment += rowset->num_segments();
+    // TODO update zonemap
+    if (rowset->rowset_meta()->has_delete_predicate()) {
+        _rowset_meta->set_delete_predicate(rowset->rowset_meta()->delete_predicate());
+    }
+    return OLAP_SUCCESS;
+}
+
+OLAPStatus BetaRowsetWriter::add_rowset_for_linked_schema_change(RowsetSharedPtr rowset,
+                                                                 const SchemaMapping& schema_mapping) {
+    // TODO use schema_mapping to transfer zonemap
+    return add_rowset(rowset);
+}
+
+OLAPStatus BetaRowsetWriter::flush() {
+    if (_segment_writer != nullptr) {
+        RETURN_NOT_OK(_flush_segment_writer());
+    }
+    return OLAP_SUCCESS;
+}
+
+RowsetSharedPtr BetaRowsetWriter::build() {
+    _rowset_meta->set_num_rows(_num_rows_written);
+    _rowset_meta->set_total_disk_size(_total_data_size);
+    _rowset_meta->set_data_disk_size(_total_data_size);
+    _rowset_meta->set_index_disk_size(0); // TODO collect index size
+    // TODO write zonemap to meta
+    _rowset_meta->set_empty(_num_rows_written == 0);
+    _rowset_meta->set_creation_time(time(nullptr));
+    _rowset_meta->set_num_segments(_num_segment);
+    if (_is_pending) {
+        _rowset_meta->set_rowset_state(COMMITTED);
+    } else {
+        _rowset_meta->set_rowset_state(VISIBLE);
+    }
+
+    RowsetSharedPtr rowset(new BetaRowset(_context.tablet_schema,
+                                          _context.rowset_path_prefix,
+                                          _context.data_dir,
+                                          _rowset_meta));
+    auto status = rowset->init();
+    if (status != OLAP_SUCCESS) {
+        LOG(WARNING) << "rowset init failed when build new rowset";
```

Comment:
add status msg to log

Reply:
Because OLAPStatus is just an enum and doesn't have errmsg field, the real reason is usually not logged by caller, but by the function who encounters error in the first place